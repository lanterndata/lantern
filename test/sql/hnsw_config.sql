\ir utils/small_world_array.sql

-- Before the HNSW index is created, the parameter hnsw.init_k should not be available
\set ON_ERROR_STOP off
SHOW hnsw.init_k;
\set ON_ERROR_STOP on

-- Create an index and verify that it was created
CREATE INDEX ON small_world USING hnsw (v) WITH (dim=3);
SELECT * FROM ldb_get_indexes('small_world');

-- Verify that hnsw.init_k exists after index creation
SHOW hnsw.init_k;

-- Modify hnsw.init_k and verify that it was modified
SET hnsw.init_k = 45;
SHOW hnsw.init_k;

-- Reset all parameters and verify that hnsw.init_k was reset
RESET ALL;
SHOW hnsw.init_k;

-- Validate the index data structures
SELECT _lantern_internal.validate_index('small_world_v_idx', false);

-- Validate that lantern.pgvector_compat disables the operator rewriting hooks
CREATE TABLE op_test (v REAL[]);
INSERT INTO op_test (v) VALUES (ARRAY[0,0,0]), (ARRAY[1,1,1]);
CREATE INDEX ON op_test USING hnsw(v dist_cos_ops);
-- should rewrite operator
SELECT * FROM op_test ORDER BY v <-> ARRAY[1,1,1];
SET lantern.pgvector_compat=TRUE;
-- should throw error
\set ON_ERROR_STOP off
SELECT * FROM op_test ORDER BY v <-> ARRAY[1,1,1];
\set ON_ERROR_STOP on
RESET ALL;
-- should rewrite operator
SELECT * FROM op_test ORDER BY v <-> ARRAY[1,1,1];
