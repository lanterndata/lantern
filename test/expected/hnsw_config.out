\ir utils/small_world_array.sql
CREATE TABLE small_world (
    id VARCHAR(3),
    b BOOLEAN,
    v REAL[3]
);
INSERT INTO small_world (id, b, v) VALUES
    ('000', TRUE,  '{0,0,0}'),
    ('001', TRUE,  '{0,0,1}'),
    ('010', FALSE, '{0,1,0}'),
    ('011', TRUE,  '{0,1,1}'),
    ('100', FALSE, '{1,0,0}'),
    ('101', FALSE, '{1,0,1}'),
    ('110', FALSE, '{1,1,0}'),
    ('111', TRUE,  '{1,1,1}');
-- Before the HNSW index is created, the parameter hnsw.init_k should not be available
\set ON_ERROR_STOP off
SHOW hnsw.init_k;
ERROR:  unrecognized configuration parameter "hnsw.init_k"
\set ON_ERROR_STOP on
-- Create an index and verify that it was created
CREATE INDEX ON small_world USING hnsw (v) WITH (dim=3);
INFO:  done init usearch index
INFO:  inserted 8 elements
INFO:  done saving 8 vectors
SELECT * FROM ldb_get_indexes('small_world');
     indexname     | size  |                                      indexdef                                      | total_index_size 
-------------------+-------+------------------------------------------------------------------------------------+------------------
 small_world_v_idx | 24 kB | CREATE INDEX small_world_v_idx ON public.small_world USING hnsw (v) WITH (dim='3') | 24 kB
(1 row)

-- Verify that hnsw.init_k exists after index creation
SHOW hnsw.init_k;
 hnsw.init_k 
-------------
 10
(1 row)

-- Modify hnsw.init_k and verify that it was modified
SET hnsw.init_k = 45;
SHOW hnsw.init_k;
 hnsw.init_k 
-------------
 45
(1 row)

-- Reset all parameters and verify that hnsw.init_k was reset
RESET ALL;
SHOW hnsw.init_k;
 hnsw.init_k 
-------------
 10
(1 row)

-- Validate the index data structures
SELECT _lantern_internal.validate_index('small_world_v_idx', false);
INFO:  validate_index() start for small_world_v_idx
INFO:  validate_index() done, no issues found.
 validate_index 
----------------
 
(1 row)

-- Validate that lantern.pgvector_compat disables the operator rewriting hooks
CREATE TABLE op_test (v REAL[]);
INSERT INTO op_test (v) VALUES (ARRAY[0,0,0]), (ARRAY[1,1,1]);
CREATE INDEX ON op_test USING hnsw(v dist_cos_ops);
INFO:  done init usearch index
INFO:  inserted 2 elements
INFO:  done saving 2 vectors
-- should rewrite operator
SELECT * FROM op_test ORDER BY v <-> ARRAY[1,1,1];
    v    
---------
 {1,1,1}
 {0,0,0}
(2 rows)

SET lantern.pgvector_compat=TRUE;
-- should throw error
\set ON_ERROR_STOP off
SELECT * FROM op_test ORDER BY v <-> ARRAY[1,1,1];
ERROR:  Operator <-> can only be used inside of an index
\set ON_ERROR_STOP on
RESET ALL;
-- should rewrite operator
SELECT * FROM op_test ORDER BY v <-> ARRAY[1,1,1];
    v    
---------
 {1,1,1}
 {0,0,0}
(2 rows)

